File: examples2d/stress_tests/ragdolls2.rs
use rapier_testbed2d::TestbedViewer;
use rapier2d::prelude::*;

/// A 10-body articulated ragdoll (torso, head, 2×2 arm links, 2×2 leg links)
/// whose torso center is at `origin`. 9 revolute joints, all with limits.
fn ragdoll(world: &mut PhysicsWorld, origin: Vec2) {
    let part = |world: &mut PhysicsWorld, offset: Vec2, collider: ColliderBuilder| {
        let body = RigidBodyBuilder::dynamic().translation(origin + offset);
        let (handle, _) = world.insert(body, collider);
        (handle, offset)
    };

    let revolute = |parent: (RigidBodyHandle, Vec2),
                    child: (RigidBodyHandle, Vec2),
                    anchor: Vec2,
                    limits: [f32; 2]| {
        RevoluteJointBuilder::new()
            .local_anchor1(anchor - parent.1)
            .local_anchor2(anchor - child.1)
            .limits(limits)
            .contacts_enabled(false)
    };

    let torso = part(world, Vec2::ZERO, ColliderBuilder::capsule_y(0.3, 0.15));
    let head = part(world, Vec2::new(0.0, 0.55), ColliderBuilder::ball(0.15));
    let neck = revolute(torso, head, Vec2::new(0.0, 0.42), [-0.5, 0.5]);
    world.insert_impulse_joint(torso.0, head.0, neck);

    for side in [-1.0f32, 1.0] {
        let upper_arm = part(
            world,
            Vec2::new(side * 0.36, 0.25),
            ColliderBuilder::capsule_x(0.14, 0.06),
        );
        let forearm = part(
            world,
            Vec2::new(side * 0.70, 0.25),
            ColliderBuilder::capsule_x(0.14, 0.06),
        );
        let thigh = part(
            world,
            Vec2::new(side * 0.09, -0.52),
            ColliderBuilder::capsule_y(0.16, 0.07),
        );
        let shin = part(
            world,
            Vec2::new(side * 0.09, -0.92),
            ColliderBuilder::capsule_y(0.16, 0.07),
        );

        let shoulder = revolute(torso, upper_arm, Vec2::new(side * 0.19, 0.25), [-1.2, 1.2]);
        world.insert_impulse_joint(torso.0, upper_arm.0, shoulder);
        let elbow = revolute(upper_arm, forearm, Vec2::new(side * 0.53, 0.25), [0.0, 2.5]);
        world.insert_impulse_joint(upper_arm.0, forearm.0, elbow);
        let hip = revolute(torso, thigh, Vec2::new(side * 0.09, -0.33), [-1.0, 1.0]);
        world.insert_impulse_joint(torso.0, thigh.0, hip);
        let knee = revolute(thigh, shin, Vec2::new(side * 0.09, -0.72), [0.0, 2.3]);
        world.insert_impulse_joint(thigh.0, shin.0, knee);
    }
}

pub async fn run(viewer: &mut TestbedViewer) -> anyhow::Result<()> {
    /*
     * World
     */
    let mut world = PhysicsWorld::new();

    /*
     * Ground
     */
    world.insert(
        RigidBodyBuilder::fixed().translation(Vec2::new(0.0, -1.0)),
        ColliderBuilder::cuboid(1000.0, 1.0),
    );

    /*
     * Ragdolls dropped into a pile: 20 columns, 10 layers (200 ragdolls,
     * 2000 dynamic bodies, 1800 limit joints).
     */
    for layer in 0..10 {
        for col in 0..20 {
            ragdoll(
                &mut world,
                Vec2::new(col as f32 * 2.2, 1.5 + layer as f32 * 2.6),
            );
        }
    }

    /*
     * Set up the testbed.
     */
    viewer.set_world(&mut world);
    viewer.look_at(Vec2::new(22.0, 6.0), 15.0);

    while viewer.render_frame(&mut world).await {
        if viewer.simulating() {
            world.step();
        }
    }
    Ok(())
}

File: c/testbed/examples2d/stress_tests/ragdolls2.c
/* Port of examples2d/stress_tests/ragdolls2.rs. */
#include "testbed.h"
#include "rapier_helpers.h"
#include "rapier_math.h"

typedef struct Part {
    R2RigidBodyHandle handle;
    R2Vector offset;
} Part;

/* A body and its offset from the ragdoll's torso. */
static Part part(R2World *world, R2Vector origin, R2Vector offset, R2ColliderDesc collider,
                 int noSleep) {
    R2RigidBodyDesc body = r2DynamicRigidBodyDesc();
    body.position.translation = r2VectorAdd(origin, offset);
    body.canSleep = !noSleep;
    Part result = {.offset = offset};
    result.handle = r2InsertRigidBody(world, &body);
    r2InsertCollider(result.handle, &collider);

    return result;
}

static R2JointDesc revolute(Part parent, Part child, R2Vector anchor, R2Real min, R2Real max) {
    R2JointDesc joint = r2RevoluteJointDesc();
    joint.localFrame1.translation = r2VectorSub(anchor, parent.offset);
    joint.localFrame2.translation = r2VectorSub(anchor, child.offset);
    r2JointDesc_SetLimits(&joint, R2_AXIS_ANG_X, min, max);
    joint.contactsEnabled = 0;
    return joint;
}

/* Ten bodies and nine limited joints, matching the Rust ragdoll. */
static void ragdoll(R2World *world, R2Vector origin, int noSleep) {
    R2ColliderDesc collider = r2CapsuleYColliderDesc(.3, .15);
    Part torso = part(world, origin, r2Vector(0, 0), collider, noSleep);
    collider = r2BallColliderDesc(.15);
    Part head = part(world, origin, r2Vector(0, 0.55), collider, noSleep);
    R2JointDesc neck = revolute(torso, head, r2Vector(0, 0.42), -.5, .5);
    R2ImpulseJointHandle jointHandle =
        r2InsertImpulseJoint(torso.handle, head.handle, &neck);

    for (int side = -1; side <= 1; side += 2) {
        collider = r2CapsuleXColliderDesc(.14, .06);
        Part upperArm = part(world, origin, r2Vector(side * .36, .25), collider, noSleep);
        collider = r2CapsuleXColliderDesc(.14, .06);
        Part forearm = part(world, origin, r2Vector(side * .70, .25), collider, noSleep);
        collider = r2CapsuleYColliderDesc(.16, .07);
        Part thigh = part(world, origin, r2Vector(side * .09, -.52), collider, noSleep);
        collider = r2CapsuleYColliderDesc(.16, .07);
        Part shin = part(world, origin, r2Vector(side * .09, -.92), collider, noSleep);
        R2JointDesc shoulder = revolute(torso, upperArm, r2Vector(side * .19, .25), -1.2, 1.2);
        jointHandle = r2InsertImpulseJoint(torso.handle, upperArm.handle, &shoulder);

        R2JointDesc elbow = revolute(upperArm, forearm, r2Vector(side * .53, .25), 0, 2.5);
        jointHandle = r2InsertImpulseJoint(upperArm.handle, forearm.handle, &elbow);

        R2JointDesc hip = revolute(torso, thigh, r2Vector(side * .09, -.33), -1, 1);
        jointHandle = r2InsertImpulseJoint(torso.handle, thigh.handle, &hip);

        R2JointDesc knee = revolute(thigh, shin, r2Vector(side * .09, -.72), 0, 2.3);
        jointHandle = r2InsertImpulseJoint(thigh.handle, shin.handle, &knee);
    }
}

void tbStressTestsRagdolls2(Testbed *testbed) {
    R2World *world = r2NewWorld();
    {
        R2RigidBodyDesc rigidBody = r2FixedRigidBodyDesc();
        rigidBody.position.translation = r2Vector(0, -1);
        rigidBody.canSleep = !testbed->noSleep;
        R2ColliderDesc collider = r2CuboidColliderDesc(r2Vector(1000, 1));
        R2RigidBodyHandle rigidBodyHandle = r2InsertRigidBody(world, &rigidBody);
        r2InsertCollider(rigidBodyHandle, &collider);
    }
    /* 200 ragdolls: 20 columns and 10 layers. */
    for (int layer = 0; layer < 10; ++layer) {
        for (int col = 0; col < 20; ++col) {
            ragdoll(world, r2Vector(col * 2.2, 1.5 + layer * 2.6), testbed->noSleep);
        }
    }
    tbCamera2(testbed, 22, 6, 15);

    tbSetWorld(testbed, world);

    while (tbRenderFrame(testbed, &world)) {
        if (tbSimulating(testbed)) {
            r2Step(world, NULL, NULL);
        }
    }
    r2FreeWorld(world);
}
